File: java/src/test/java/io/vedadb/wire/vbp/VBPTypeCodecTest.java
package io.vedadb.wire.vbp;

import org.junit.jupiter.api.Test;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.*;

class VBPTypeCodecTest {

    @Test
    void encodeBoolTrue() {
        assertArrayEquals(new byte[]{1}, VBPTypeCodec.encodeBool(true));
    }

    @Test
    void encodeBoolFalse() {
        assertArrayEquals(new byte[]{0}, VBPTypeCodec.encodeBool(false));
    }

    @Test
    void encodeInt2() {
        assertArrayEquals(new byte[]{(byte) 0xFF, (byte) 0xFF}, VBPTypeCodec.encodeInt2((short) -1));
        assertArrayEquals(new byte[]{0x00, 0x7F}, VBPTypeCodec.encodeInt2((short) 32512));
    }

    @Test
    void encodeInt4LittleEndian() {
        assertArrayEquals(new byte[]{0x78, 0x56, 0x34, 0x12}, VBPTypeCodec.encodeInt4(0x12345678));
    }

    @Test
    void encodeInt8LittleEndian() {
        // 0x123456789ABCDEF0 little-endian: F0 DE BC 9A 78 56 34 12
        byte[] b = VBPTypeCodec.encodeInt8(0x123456789ABCDEF0L);
        assertEquals(8, b.length);
        assertEquals((byte) 0xF0, b[0]);
        assertEquals((byte) 0xDE, b[1]);
        assertEquals((byte) 0xBC, b[2]);
        assertEquals((byte) 0x9A, b[3]);
        assertEquals((byte) 0x78, b[4]);
        assertEquals((byte) 0x56, b[5]);
        assertEquals((byte) 0x34, b[6]);
        assertEquals((byte) 0x12, b[7]);
    }

    @Test
    void encodeDecodeRoundTripInt4() {
        byte[] enc = VBPTypeCodec.encodeInt4(0x12345678);
        assertEquals(0x12345678, VBPTypeCodec.decodeInt4(enc));
    }

    @Test
    void encodeDecodeRoundTripInt8() {
        byte[] enc = VBPTypeCodec.encodeInt8(-1L);
        assertEquals(-1L, VBPTypeCodec.decodeInt8(enc));
    }

    @Test
    void encodeFloat8Pi() {
        byte[] b = VBPTypeCodec.encodeFloat8(Math.PI);
        assertEquals(Math.PI, VBPTypeCodec.decodeFloat8(b), 1e-15);
    }

    @Test
    void encodeFloat4() {
        byte[] b = VBPTypeCodec.encodeFloat4(1.5f);
        assertEquals(1.5f, VBPTypeCodec.decodeFloat4(b), 1e-6);
    }

    @Test
    void encodeTextIsLengthPrefixed() {
        byte[] b = VBPTypeCodec.encodeText("hi");
        assertEquals(4 + 2, b.length);
        assertEquals(2, b[0]); // length
        assertEquals('h', b[4]);
        assertEquals('i', b[5]);
    }

    @Test
    void encodeDecodeText() {
        assertEquals("hello", VBPTypeCodec.decodeText(VBPTypeCodec.encodeText("hello")));
    }

    @Test
    void encodeEmptyText() {
        byte[] b = VBPTypeCodec.encodeText("");
        assertEquals(4, b.length);
        assertEquals(0, b[0]);
    }

    @Test
    void encodeByteaRoundTrip() {
        byte[] raw = {1, 2, 3, 4, 5};
        assertArrayEquals(raw, VBPTypeCodec.decodeBytea(VBPTypeCodec.encodeBytea(raw)));
    }

    @Test
    void encodeUuidIsSixteenBytes() {
        UUID u = UUID.fromString("12345678-1234-5678-1234-567812345678");
        byte[] b = VBPTypeCodec.encodeUuid(u);
        assertEquals(16, b.length);
        assertEquals(u, VBPTypeCodec.decodeUuid(b));
    }

    @Test
    void encodeTimestamp() {
        byte[] b = VBPTypeCodec.encodeTimestamp(0L);
        assertEquals(8, b.length);
        assertEquals(0L, VBPTypeCodec.decodeTimestamp(b));
    }

    @Test
    void encodeInterval() {
        byte[] b = VBPTypeCodec.encodeInterval(123_456L, 7, 3);
        assertEquals(16, b.length);
    }

    @Test
    void encodeMoney() {
        byte[] b = VBPTypeCodec.encodeMoney(100L);
        assertEquals(8, b.length);
    }

    @Test
    void inputEnvelopeNull() {
        byte[] env = VBPTypeCodec.inputEnvelopeNull(VBPOpcodes.T_INT4);
        assertEquals(3, env.length);
        assertEquals(1, env[0]); // NULL_TAG
        assertEquals(23, env[1] & 0xFF);
        assertEquals(0, env[2] & 0xFF);
    }

    @Test
    void inputEnvelopeNotNull() {
        byte[] body = VBPTypeCodec.encodeInt4(42);
        byte[] env = VBPTypeCodec.inputEnvelope(VBPOpcodes.T_INT4, body);
        assertEquals(3 + 4 + body.length, env.length);
        assertEquals(0, env[0]); // NOT_NULL tag
        assertEquals(23, env[1] & 0xFF);
    }

    @Test
    void outputEnvelopeRow() {
        byte[] body = VBPTypeCodec.encodeInt4(99);
        byte[] env = VBPTypeCodec.outputEnvelopeRow(new int[]{VBPOpcodes.T_INT4}, new byte[][]{body}, (byte) 0);
        assertTrue(env.length >= 4 + 4 + 2 + 1 + body.length);
    }

    @Test
    void selectOneRowIsValid() {
        byte[] b = VBPTypeCodec.selectOneRow();
        VBPTypeCodec.DataChunk dc = VBPTypeCodec.parseDataChunk(b);
        assertEquals(1, dc.nColumns);
        assertEquals(1, dc.rowValues.size());
        assertEquals(1, dc.rowValues.get(0));
    }

    @Test
    void rowsFinishedBody() {
        byte[] b = VBPTypeCodec.rowsFinished(5, 2, new int[]{23, 25});
        assertEquals(4 + 4 + 4, b.length);
    }

    @Test
    void commandCompleteBody() {
        byte[] b = VBPTypeCodec.commandComplete("INSERT 0 1", 1);
        assertEquals(4 + "INSERT 0 1".length() + 8, b.length);
    }

    @Test
    void errorBodyRoundTrip() {
        byte[] b = VBPTypeCodec.errorBody("42P01", "undefined table", "no detail", "use CREATE");
        VBPTypeCodec.ErrorParts ep = VBPTypeCodec.parseErrorBody(b);
        assertEquals("42P01", ep.sqlstate);
        assertEquals("undefined table", ep.message);
        assertEquals("no detail", ep.detail);
        assertEquals("use CREATE", ep.hint);
    }

    @Test
    void clientHelloBodyParsesBack() {
        byte[] b = VBPTypeCodec.clientHelloBody(1, 0, "admin", "main", (byte) 0, "admin");
        // Length: 2+2+4+5+4+4+1+4+5 = 31
        assertEquals(31, b.length);
    }

    @Test
    void serverReadyParse() {
        java.nio.ByteBuffer bb = java.nio.ByteBuffer.allocate(13).order(java.nio.ByteOrder.LITTLE_ENDIAN);
        bb.putInt(0x000A0000); // v10.0.0
        bb.putInt(0x0000001F); // caps
        bb.put((byte) 0); // auth_required
        bb.putInt(0); // nonce_len
        VBPTypeCodec.ServerReadyParts sr = VBPTypeCodec.parseServerReady(bb.array());
        assertEquals(0x000A0000, sr.serverVersion);
        assertEquals(0x0000001F, sr.serverCaps);
        assertFalse(sr.authRequired);
    }

    @Test
    void serverReadyParseAuthRequired() {
        java.nio.ByteBuffer bb = java.nio.ByteBuffer.allocate(21).order(java.nio.ByteOrder.LITTLE_ENDIAN);
        bb.putInt(0x000A0000);
        bb.putInt(0x0000001F);
        bb.put((byte) 1); // auth_required
        bb.putInt(8);
        byte[] nonce = {1,2,3,4,5,6,7,8};
        bb.put(nonce);
        VBPTypeCodec.ServerReadyParts sr = VBPTypeCodec.parseServerReady(bb.array());
        assertTrue(sr.authRequired);
        assertEquals(8, sr.nonce.length);
    }

    @Test
    void authOkParse() {
        java.nio.ByteBuffer bb = java.nio.ByteBuffer.allocate(20).order(java.nio.ByteOrder.LITTLE_ENDIAN);
        bb.putLong(0xC0FFEE);
        bb.putLong(0x123456789ABCDEF0L);
        bb.putInt(0);
        VBPTypeCodec.AuthOkParts ok = VBPTypeCodec.parseAuthOk(bb.array());
        assertEquals(0xC0FFEE, ok.sessionTokenLo);
        assertEquals(0x123456789ABCDEF0L, ok.sessionTokenHi);
    }

    @Test
    void queryBodyParse() {
        byte[] body = VBPTypeCodec.queryBody(7, "SELECT 1", java.util.Collections.emptyList());
        VBPTypeCodec.QueryParts qp = VBPTypeCodec.parseQuery(body);
        assertEquals(7, qp.queryId);
        assertEquals("SELECT 1", qp.sql);
        assertEquals(0, qp.nParams);
    }
}

File: php/tests/Wire/Vbp/VBPTypeCodecTest.php
<?php
declare(strict_types=1);

namespace VedaDB\Tests\Wire\Vbp;

use PHPUnit\Framework\TestCase;
use VedaDB\Wire\Vbp\VBPTypeCodec;
use VedaDB\Wire\Vbp\VBPOpcodes as Ops;

final class VBPTypeCodecTest extends TestCase
{
    // -- Fixed-width encoders --

    public function testEncodeBool(): void
    {
        $this->assertSame("\x01", VBPTypeCodec::encodeBool(true));
        $this->assertSame("\x00", VBPTypeCodec::encodeBool(false));
    }

    public function testEncodeInt2(): void
    {
        $this->assertSame("\x2a\x00", VBPTypeCodec::encodeInt2(42));
        $this->assertSame("\xff\xff", VBPTypeCodec::encodeInt2(-1));
    }

    public function testEncodeInt4(): void
    {
        $this->assertSame(pack('V', 42), VBPTypeCodec::encodeInt4(42));
    }

    public function testEncodeInt8(): void
    {
        $this->assertSame(pack('P', 100), VBPTypeCodec::encodeInt8(100));
    }

    public function testEncodeFloat4(): void
    {
        $b = VBPTypeCodec::encodeFloat4(1.5);
        $this->assertSame(4, strlen($b));
        $this->assertEqualsWithDelta(1.5, VBPTypeCodec::decodeFloat4($b), 0.0001);
    }

    public function testEncodeFloat8(): void
    {
        $b = VBPTypeCodec::encodeFloat8(2.5);
        $this->assertSame(8, strlen($b));
        $this->assertEqualsWithDelta(2.5, VBPTypeCodec::decodeFloat8($b), 0.0001);
    }

    public function testEncodeTextIsLengthPrefixed(): void
    {
        $b = VBPTypeCodec::encodeText("hi");
        $this->assertSame(pack('V', 2) . "hi", $b);
    }

    public function testEncodeVarchar(): void
    {
        $b = VBPTypeCodec::encodeVarchar("test");
        $this->assertSame(pack('V', 4) . "test", $b);
    }

    public function testEncodeByteaIsLengthPrefixed(): void
    {
        $b = VBPTypeCodec::encodeBytea("\x01\x02");
        $this->assertSame(pack('V', 2) . "\x01\x02", $b);
    }

    public function testEncodeUuidAcceptsCanonicalAndHex(): void
    {
        $uuid = '12345678-1234-5678-1234-567812345678';
        $hex = '12345678123456781234567812345678';
        $this->assertSame(hex2bin($hex), VBPTypeCodec::encodeUuid($uuid));
        $this->assertSame(hex2bin($hex), VBPTypeCodec::encodeUuid($hex));
    }

    public function testEncodeUuidInvalidThrows(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        VBPTypeCodec::encodeUuid('not-a-uuid');
    }

    public function testEncodeJsonb(): void
    {
        $b = VBPTypeCodec::encodeJsonb('{"a":1}');
        $this->assertSame(pack('V', 7) . '{"a":1}', $b);
    }

    public function testEncodeNumericIsLengthPrefixed(): void
    {
        $b = VBPTypeCodec::encodeNumeric("123.45");
        $this->assertSame(pack('V', 6) . "123.45", $b);
    }

    public function testEncodeDate(): void
    {
        $this->assertSame(pack('V', 19000), VBPTypeCodec::encodeDate(19000));
    }

    public function testEncodeMoney(): void
    {
        $this->assertSame(pack('P', 100), VBPTypeCodec::encodeMoney(100));
    }

    // -- Decoders --

    public function testDecodeBool(): void
    {
        $this->assertTrue(VBPTypeCodec::decodeBool("\x01"));
        $this->assertFalse(VBPTypeCodec::decodeBool("\x00"));
    }

    public function testDecodeInt4(): void
    {
        $this->assertSame(42, VBPTypeCodec::decodeInt4(pack('V', 42)));
    }

    public function testDecodeInt8(): void
    {
        $this->assertSame(12345, VBPTypeCodec::decodeInt8(pack('P', 12345)));
    }

    public function testDecodeInt2(): void
    {
        $this->assertSame(0x7fff, VBPTypeCodec::decodeInt2(pack('v', 0x7fff)));
        $this->assertSame(-1, VBPTypeCodec::decodeInt2(pack('v', 0xffff)));
    }

    public function testDecodeText(): void
    {
        $b = pack('V', 5) . "hello";
        $this->assertSame("hello", VBPTypeCodec::decodeText($b));
    }

    public function testDecodeUuid(): void
    {
        $hex = '12345678123456781234567812345678';
        $this->assertSame(
            '12345678-1234-5678-1234-567812345678',
            VBPTypeCodec::decodeUuid(hex2bin($hex)),
        );
    }

    // -- Input envelope --

    public function testInputEnvelopeNonNull(): void
    {
        $env = VBPTypeCodec::inputEnvelope(Ops::T_INT4, pack('V', 7));
        // NULL_TAG_NOT_NULL + u16 typeId + u32 len + body
        $this->assertSame(
            chr(0) . pack('v', Ops::T_INT4) . pack('V', 4) . pack('V', 7),
            $env,
        );
    }

    public function testInputEnvelopeNull(): void
    {
        $env = VBPTypeCodec::inputEnvelopeNull(Ops::T_INT4);
        $this->assertSame(chr(1) . pack('v', Ops::T_INT4), $env);
    }

    // -- Output envelope --

    public function testOutputEnvelopeRowSingleInt4(): void
    {
        $env = VBPTypeCodec::outputEnvelopeRow([Ops::T_INT4], [pack('V', 1)]);
        $decoded = VBPTypeCodec::parseDataChunk($env);
        $this->assertSame(1, $decoded['nColumns']);
        $this->assertSame([Ops::T_INT4], $decoded['colTypes']);
        $this->assertSame([1], $decoded['values']);
    }

    public function testOutputEnvelopeRowMultiColumn(): void
    {
        $env = VBPTypeCodec::outputEnvelopeRow(
            [Ops::T_INT4, Ops::T_TEXT, Ops::T_BOOL],
            [pack('V', 99), pack('V', 3) . "foo", "\x01"],
        );
        $decoded = VBPTypeCodec::parseDataChunk($env);
        $this->assertSame(3, $decoded['nColumns']);
        $this->assertSame([99, "foo", true], $decoded['values']);
    }

    public function testOutputEnvelopeRowWithNull(): void
    {
        $env = VBPTypeCodec::outputEnvelopeRow(
            [Ops::T_INT4, Ops::T_TEXT],
            [null, pack('V', 3) . "bar"],
            0b01, // bit 0 = first col NULL
        );
        $decoded = VBPTypeCodec::parseDataChunk($env);
        $this->assertNull($decoded['values'][0]);
        $this->assertSame("bar", $decoded['values'][1]);
    }

    public function testSelectOneRow(): void
    {
        $env = VBPTypeCodec::selectOneRow();
        $decoded = VBPTypeCodec::parseDataChunk($env);
        $this->assertSame([1], $decoded['values']);
    }

    // -- ROWS_FINISHED --

    public function testRowsFinishedBody(): void
    {
        $body = VBPTypeCodec::rowsFinished(3, 2, [Ops::T_INT4, Ops::T_TEXT]);
        $u = unpack('VnRows/VnCols', substr($body, 0, 8));
        $this->assertSame(3, $u['nRows']);
        $this->assertSame(2, $u['nCols']);
    }

    // -- COMMAND_COMPLETE --

    public function testCommandCompleteBody(): void
    {
        $body = VBPTypeCodec::commandComplete("SELECT", 5);
        $u = unpack('VtLen', substr($body, 0, 4));
        $this->assertSame(6, $u['tLen']);
        $this->assertSame("SELECT", substr($body, 4, 6));
    }

    // -- ERROR body --

    public function testErrorBody(): void
    {
        $body = VBPTypeCodec::errorBody("42601", "syntax error", "at line 1", "check quotes");
        $parsed = VBPTypeCodec::parseErrorBody($body);
        $this->assertSame("42601", $parsed['sqlstate']);
        $this->assertSame("syntax error", $parsed['message']);
        $this->assertSame("at line 1", $parsed['detail']);
        $this->assertSame("check quotes", $parsed['hint']);
    }

    public function testErrorBodyPaddedSqlstate(): void
    {
        $body = VBPTypeCodec::errorBody("42", "short");
        $parsed = VBPTypeCodec::parseErrorBody($body);
        $this->assertSame("42000", $parsed['sqlstate']);
    }

    // -- CLIENT_HELLO --

    public function testClientHelloBody(): void
    {
        $body = VBPTypeCodec::clientHelloBody(1, 0, "admin", "main", 0, "admin");
        $this->assertGreaterThan(0, strlen($body));
        $u = unpack('vver/vflags', substr($body, 0, 4));
        $this->assertSame(1, $u['ver']);
        $this->assertSame(0, $u['flags']);
    }

    // -- SERVER_READY parser --

    public function testParseServerReady(): void
    {
        $body = pack('VV', 100, 0x42) . "\x01" . pack('V', 4) . "ABCD";
        $sr = VBPTypeCodec::parseServerReady($body);
        $this->assertSame(100, $sr['serverVersion']);
        $this->assertSame(0x42, $sr['serverCaps']);
        $this->assertTrue($sr['authRequired']);
        $this->assertSame("ABCD", $sr['nonce']);
    }

    // -- AUTH_OK parser --

    public function testParseAuthOk(): void
    {
        $body = pack('P', 12345) . pack('P', 67890) . pack('V', 7);
        $parsed = VBPTypeCodec::parseAuthOk($body);
        $this->assertSame(12345, $parsed['sessionTokenLo']);
        $this->assertSame(67890, $parsed['sessionTokenHi']);
        $this->assertSame(7, $parsed['serverFlags']);
    }

    // -- QUERY body --

    public function testQueryBodyNoParams(): void
    {
        $body = VBPTypeCodec::queryBody(1, "SELECT 1");
        $parsed = VBPTypeCodec::parseQuery($body);
        $this->assertSame(1, $parsed['queryId']);
        $this->assertSame("SELECT 1", $parsed['sql']);
        $this->assertSame(0, $parsed['nParams']);
    }

    public function testQueryBodyWithParams(): void
    {
        $env = VBPTypeCodec::inputEnvelope(Ops::T_INT4, pack('V', 7));
        $body = VBPTypeCodec::queryBody(2, "SELECT ?", [$env]);
        $parsed = VBPTypeCodec::parseQuery($body);
        $this->assertSame(2, $parsed['queryId']);
        $this->assertSame("SELECT ?", $parsed['sql']);
        $this->assertSame(1, $parsed['nParams']);
    }

    // -- PING body --

    public function testPingBody(): void
    {
        $nonce = random_bytes(8);
        $body = VBPTypeCodec::pingBody($nonce);
        $this->assertSame(8, strlen($body));
        $this->assertSame($nonce, $body);
    }

    public function testPingBodyInvalidLengthThrows(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        VBPTypeCodec::pingBody("abc");
    }
}
